include/ts/nexthop.h

```diff
@@ -28,22 +28,28 @@
 
 #pragma once
 
-#include <ts/apidefs.h>
-
-// plugin callback commands.
-enum NHCmd { NH_MARK_UP, NH_MARK_DOWN };
-
-struct NHHealthStatus {
-  virtual bool isNextHopAvailable(TSHttpTxn txn, const char *hostname, const int port, void *ih = nullptr) = 0;
-  virtual void markNextHop(TSHttpTxn txn, const char *hostname, const int port, const NHCmd status, void *ih = nullptr,
-                           const time_t now = 0)                                                           = 0;
-  virtual ~NHHealthStatus() {}
-};
-
-struct NHPluginStrategy {
-  virtual void findNextHop(TSHttpTxn txnp, void *ih = nullptr)   = 0;
-  virtual bool nextHopExists(TSHttpTxn txnp, void *ih = nullptr) = 0;
-  virtual ~NHPluginStrategy() {}
-
-  NHHealthStatus *healthStatus;
-};
+// Plugins may set this to indicate how to retry.
+//
+// If handled is false, then no plugin set it, and Core will proceed to do its own thing.
+//
+// If handled is true, core will not do any parent processing, markdown, or anything else,
+// but will use the values in this for whether to use the existing response or make another request,
+// and what that request should look like.
+//
+// See the API functions which take this for ownership requirements of pointers, like hostname.
+//
+// DO NOT put non-POD data in this. It's passed by-copy from API calls.
+//
+typedef struct {
+  // TODO this shouldn't be necessary - plugins should manipulate the response as they see fit,
+  // core shouldn't "know" if it was a "success" or "failure," only the response or retry data/action.
+  // But for now, core needs to know, for reasons.
+  bool fail;
+  const char *hostname;
```

Comment:
Add `length` - don't require a C-string.

Reply:
Should I comment/document that the hostname must still be null-terminated?
`std::string_view` isn't necessarily null-terminated, but in core this hostname is set to `s->parent_result.hostname` and all of core is expecting a null-terminated string. Allowing non-null-terminated `respose_action.hostname` would require a lot of work and checking, and would be very difficult to be sure it wasn't missed somewhere.

Reply:
Gah. That defeats the point of having that API. I guess change it for now and put a note and I'll need to fix the internals at some later point.

Reply:
Yeah, that char* is passed around pretty carelessly in core, unfortunately.
Well, if we document it as requiring a null terminator for now, we can remove that requirement later, after we've cleaned up core.

Reply:
Changed to take a length